### DISCUSSION TITLE
MUI read only fields

### DISCUSSION THREAD

**Discussion by Cat30Mk3** [olikraus/u8g2#2312] [Q&A] [answered]
Hi Oliver,
I've just started using the MUI library and have poured over all the documentation and examples.
However I haven't found a way to include a READ ONLY numeric field into a form from a global integer. The global integer will be changing values via an interrupt, or through the loop while the menu is active.

Your MUICountDown example uses the mui_u8g2_u8_min_max_wm_mud_pi callback for numeric data  but this is an INPUT field that enables editing. The same applies to the bar graph.

The demo example suggests that the generalized  MUIF macro MUIF(id,cflags,data,cb) cflag can mark the field as as editable. So perhaps this could be used to prevent cursor selection of an input field for read only purposes? But I don't understand how I would use that macro. Or perhaps there is another method?

I experimented with MUI_LABEL(x,y,text) but it doesn't accept a string pointer.

I feel I'm missing something very basic. Can you offer some suggestions?

Thanks, John

**Comment by olikraus** [comment]
Hmm good question... 
Did you use MUIF_VARIABLE() to make the int variable visible?
Instead of 
https://github.com/olikraus/u8g2/blob/34be7c84883a3ffc927e04eda7a99c539da040ee/csrc/mui.h#L217
you could create a new MUIF:

``` C++
#define MUIF_READ_ONLY_VARIABLE(id,var,cb) MUIF(id,0,(var),cb)
```
Just place this macro at the beginning of your .ino file (after includes).
Then use MUIF_READ_ONLY_VARIABLE instead of MUIF_VARIABLE

Not tested, but I assume it should work.

**Reply by Cat30Mk3** [reply]
This is a great explanation of how to DIY a macro with MUIF based on yours, but as far as I can tell MUIF_VARIABLE doesn't have callbacks for number displays .. just checkbox, radio, combo and dropdown. So I used your MUIF_U8G2_U8_MIN_MAX suggestion.

**Comment by olikraus** [comment, marked as answer]
Of course you can also make a copy of this macro:
https://github.com/olikraus/u8g2/blob/34be7c84883a3ffc927e04eda7a99c539da040ee/csrc/mui_u8g2.h#L243-L246
by using 0 instead of "MUIF_CFLAG_IS_CURSOR_SELECTABLE"

It could be then used to display the value as a bar graph like demonstrated here: https://github.com/olikraus/u8g2/wiki/muiref#bar-graph-uint8

**Reply by Cat30Mk3** [reply]
This works! I was using MUIF_U8G2_U8_MIN_MAX to display the global variable so I  created 
#define MUIF_READ_ONLY_U8_MIN_MAX(id, valptr, min, max, muif) 
MUIF(id, 0, (void *)((mui_u8g2_u8_min_max_t [] ) 
{{ (valptr) MUI_U8G2_COMMA (min) MUI_U8G2_COMMA (max)}}),(muif))
and now the cursor skips this field .. just what i needed. And now I know how to create my own macros based on yours! Thanks very much!
Now I'm struggling with the fact that the global variable is uint16_t .. but I'll try some ideas before asking for help.  Thanks again. John

**Reply by olikraus** [reply]
The variable must be uint8_t because only the address of that function is transfered to the render procedures. Using uint16_t would require to create new render procedures.

You can see the problem in line 703:
https://github.com/olikraus/u8g2/blob/34be7c84883a3ffc927e04eda7a99c539da040ee/csrc/mui_u8g2.c#L697-L706

All calls to "mui_u8g2_u8mm_get_valptr(vmm)"  have to be replaced with a call to a u16 counterpart... and there are a lot of calls.

And of course we would need a new structure for this:
https://github.com/olikraus/u8g2/blob/34be7c84883a3ffc927e04eda7a99c539da040ee/csrc/mui_u8g2.h#L106-L111

**Reply by Cat30Mk3** [reply]
Ok, thanks. This confirms my approach is right. Just to prove it works I hacked up a copy of mui_u8g2.h changing the relevant uint8_t to uint16_t, adjusting buf and cnt in mui_u8g2_u8_vmm_draw_wm_pi  as well as replacing u8x8_u8toa with u8x8_u16toa. And it works! I now have  4 digit global integers displayed, some editable and some read only. Of course Ive broken the examples that use uint8_t - i need to make my solution separate without touching your code if I can.

I'm quite liking your mui project at this point. I truly appreciate your effort and quick responses. John
